tests/linearizability/id.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linearizability
+
+import "sync/atomic"
+
+type idProvider interface {
+	ClientId() int
+	RequestId() int
+}
+
+func newIdProvider() idProvider {
+	return &atomicProvider{}
+}
+
+type atomicProvider struct {
+	clientId  atomic.Int64
+	requestId atomic.Int64
+}
+
+func (id *atomicProvider) ClientId() int {
+	// Substract one as ClientId should start from zero.
+	return int(id.clientId.Add(1) - 1)
+}
+
+func (id *atomicProvider) RequestId() int {
+	return int(id.requestId.Add(1))
+}
```

Comment:
Suggest to rewrite the idProvider something like below. It shouldn't care about the business logic (e.g. `clientId`, `requestId`). 

```
type idProvider interface {
	NextId() int64
}

func newIdProvider(startFromZero bool) idProvider {
	idp := &atomicProvider{}
	if startFromZero {
		idp.nextId.Add(-1)
	}
	return idp
}

type atomicProvider struct {
	startFromZero bool
	nextId        atomic.Int64
}

func (idp *atomicProvider) NextId() int64 {
	return idp.nextId.Add(1)
}
```

Reply:
It's OK to refactor this in a separate PR.

Reply:
Proposed refactor doesn't benefit current code. It could help with code reuse, but we are not planning any reuse here. 

My main motivation to refactor this code to idProvider was to avoid writing to many comments explaining why ids behave as they do. By encapsulating logic we can make it easier to understand.